### PR TITLE
Optimize node startup process & Normalize system.properties management

### PR DESCRIPTION
**Pull request by liyuheng55555**
- Delay the system.properties generation. Now in most cases it's not necessary to manually delete data dir after startup fail.
- Use SystemPropertiesFileHandler to do every operation.

### CODE REVIEW

**Review by liyuheng55555 — COMMENTED** [review]
TBD

**Review by liyuheng55555 — COMMENTED** [review]
LGTM

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+import java.util.Properties;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class SystemPropertiesHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SystemPropertiesHandler.class);
+
+  private final File formalFile;
+  private final File tmpFile;
+  //    private ImmutableProperties cache;
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java:46]: 。

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java [hunk 2]
```diff
@@ -170,9 +167,6 @@ private DataNode() {
 
   // TODO: This needs removal of statics ...
   public static void reinitializeStatics() {
-    SYSTEM_PROPERTIES =
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java:173]: add init

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java [hunk 3]
```diff
@@ -397,15 +386,18 @@ private void storeRuntimeConfigurations(
   /**
    * Register this DataNode into cluster.
    *
+   * @param isPreCheck do pre-check before formal registration
    * @throws StartupException if register failed.
-   * @throws IOException if serialize cluster name and dataNode Id failed.
+   * @throws IOException if serialize cluster name and DataNode id failed.
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java:391]: datanode

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/service/ConfigNode.java [hunk 4]
```diff
@@ -205,7 +205,9 @@ public void active() {
       // We set up Non-Seed ConfigNode's RPC service before sending the register request
       // in order to facilitate the scheduling of capacity expansion process in ConfigNode-leader
       setUpRPCService();
-      sendRegisterConfigNodeRequest();
+      sendRegisterConfigNodeRequest(true);
+      SystemPropertiesUtils.storeSystemParameters();
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/service/ConfigNode.java:209]: remove

File: iotdb-protocol/thrift-confignode/src/main/thrift/confignode.thrift [hunk 5]
```diff
@@ -404,6 +405,7 @@ struct TConfigNodeRegisterReq {
   1: required TClusterParameters clusterParameters
   2: required common.TConfigNodeLocation configNodeLocation
   3: optional TNodeVersionInfo versionInfo
+  4: optional bool preCheck
```
Comment (OneSizeFitsQuorum) [inline iotdb-protocol/thrift-confignode/src/main/thrift/confignode.thrift:408]: remove

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 6]
```diff
@@ -407,11 +407,12 @@ public DataSet registerDataNode(TDataNodeRegisterReq req) {
               req.getClusterName(),
               req.getDataNodeConfiguration().getLocation(),
               this);
-      if (status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+      if (!req.isPreCheck() && status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
         return nodeManager.registerDataNode(req);
       }
     }
     DataNodeRegisterResp resp = new DataNodeRegisterResp();
+    resp.setPreCheck(req.isPreCheck());
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:415]: do not add this

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/datanode/DataNodeRegisterResp.java [hunk 7]
```diff
@@ -84,7 +88,7 @@ public TDataNodeRegisterResp convertToRpcDataNodeRegisterResp() {
     resp.setStatus(status);
     resp.setConfigNodeList(configNodeList);
 
-    if (status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+    if (!preCheck && status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/datanode/DataNodeRegisterResp.java:91]: .

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/datanode/DataNodeRegisterResp.java [hunk 8]
```diff
@@ -38,8 +38,8 @@ public class DataNodeRegisterResp implements DataSet {
   private TSStatus status;
   private List<TConfigNodeLocation> configNodeList;
   private Integer dataNodeId;
-
   private TRuntimeConfiguration runtimeConfiguration;
+  private boolean preCheck = false;
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/datanode/DataNodeRegisterResp.java:42]: .

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/conf/SystemPropertiesUtils.java [hunk 9]
```diff
@@ -74,11 +66,7 @@ private SystemPropertiesUtils() {
 
   // TODO: This needs removal of statics ...
   public static void reinitializeStatics() {
-    systemPropertiesFile =
-        new File(
-            ConfigNodeDescriptor.getInstance().getConf().getSystemDir()
-                + File.separator
-                + ConfigNodeConstant.SYSTEM_FILE_NAME);
+    systemPropertiesHandler = SystemPropertiesHandler.getInstance();
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/conf/SystemPropertiesUtils.java:69]: init

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  public static void init(String filePath) throws StartupException {
+    Holder.INSTANCE = new SystemPropertiesHandler(filePath);
+    Holder.INSTANCE.recover();
+  }
+
+  public static SystemPropertiesHandler getInstance() {
+    if (Holder.INSTANCE == null) {
+      LOGGER.warn(
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java:180]: printstacktrace

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    if (!tmpFile.renameTo(formalFile)) {
+      String msg =
+          String.format(
+              "Failed to replace formal system properties file, you may manually rename it: %s -> %s",
+              tmpFile.getAbsolutePath(), formalFile.getAbsolutePath());
+      throw new IOException(msg);
+    }
+  }
+
+  public static void init(String filePath) throws StartupException {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/file/SystemPropertiesHandler.java:173]: try to remove filepath

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeSystemPropertiesHandler.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+public class ConfigNodeSystemPropertiesHandler extends SystemPropertiesHandler {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(ConfigNodeSystemPropertiesHandler.class);
+
+  private ConfigNodeSystemPropertiesHandler(String filePath) {
+    super(filePath);
+  }
+
+  public static SystemPropertiesHandler getInstance() {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/conf/ConfigNodeSystemPropertiesHandler.java:37]: lock？

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/DataNodeSystemPropertiesHandler.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+public class DataNodeSystemPropertiesHandler extends SystemPropertiesHandler {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(DataNodeSystemPropertiesHandler.class);
+
+  private DataNodeSystemPropertiesHandler(String filePath) {
+    super(filePath);
+  }
+
+  public static SystemPropertiesHandler getInstance() {
+    if (Holder.INSTANCE == null) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/DataNodeSystemPropertiesHandler.java:40]: lock？

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**